File: tools/src/data/collection.ts
/**
 * A queryable, iterable view over one entity collection.
 *
 * Indexes (by id, by normalized name, by faction) are built once at construction.
 * Records are deduplicated by {@link CollectionConfig.dedupeKeyOf} (default: id,
 * first occurrence wins). Some records are intentionally shared: the same id
 * (e.g. unit `ministorum-priest`, ability `deadly-demise-d3`) appears under
 * several factions with per-faction copies that may diverge, so those
 * collections dedupe on `(faction_id, id)` to keep each faction's copy and
 * resolve faction-scoped.
 *
 * `get(id)`/`find` return the first match when an id is shared across factions;
 * use {@link Collection.byFaction} or {@link Collection.findAll} to disambiguate.
 *
 * Collections whose copies genuinely diverge per faction (units: different
 * points, keywords and profiles) set {@link CollectionConfig.guardUnscoped} so a
 * faction-less {@link Collection.get} of an id that exists under several factions
 * throws outside production — a faction-unaware lookup there is a bug (it would
 * silently return whichever faction's copy registered first). Code that has no
 * faction context on purpose (roster import, the conformance runner) calls
 * {@link Collection.getAny} instead.
 *
 * @packageDocumentation
 */
import { externalRefKey } from "../external-refs.js";
import { normalizeName } from "./normalize.js";

/**
 * True only in a production build. The {@link CollectionConfig.guardUnscoped}
 * tripwire throws everywhere else (dev servers, `vitest`, CLI tools) so an
 * ambiguous unscoped lookup surfaces immediately; in production it degrades to
 * the historical first-wins behaviour rather than crashing a user's session.
 */
const PRODUCTION =
  typeof process !== "undefined" && process.env?.NODE_ENV === "production";

/** How a {@link Collection} reads keys and builds views from raw records. */
export interface CollectionConfig<T, V> {
  items: T[];
  /** Primary id of a record (e.g. `u => u.id`, `a => a.ability_id`). */
  idOf: (item: T) => string;
  /**
   * Uniqueness key used for deduplication. Defaults to {@link idOf}. Set to a
   * composite (e.g. `(faction_id, id)`) for records that share an id across
   * factions, so distinct copies are preserved rather than collapsed.
   */
  dedupeKeyOf?: (item: T) => string;
  /** Display name, if the record has one — drives {@link Collection.find}. */
  nameOf?: (item: T) => string | undefined;
  /**
   * Alternate names a record answers to (spelling variants from other tools'
   * exports). Indexed alongside {@link nameOf} so {@link Collection.find} /
   * {@link Collection.findAll} match an alias exactly, but never returned as the
   * record's display name. The canonical name always wins a collision.
   */
  aliasesOf?: (item: T) => readonly string[] | null | undefined;
  /**
   * Renamed-id map (old id → current id), consulted by id lookups
   * ({@link Collection.get}/{@link Collection.getAny}/{@link Collection.getInFaction}/
   * {@link Collection.has}) only when the exact id misses. Lets a persisted
   * reference to a since-renamed id (e.g. a saved roster or share link authored
   * before an enhancement id was normalized) still resolve to the current
   * record. The map is expected to be pre-flattened (each key maps directly to a
   * terminal live id), so a single hop suffices. Typically the share registry's
   * `aliases`.
   */
  idAliases?: Readonly<Record<string, string>>;
  /** Stable source identities indexed by {@link Collection.byExternalRef}. */
  externalRefsOf?: (
    item: T,
  ) => readonly { namespace: string; id: string }[] | null | undefined;
  /** Owning faction id, if applicable — drives {@link Collection.byFaction}. */
  factionOf?: (item: T) => string | null | undefined;
  /**
   * When set, a faction-less {@link Collection.get} of an id that exists under
   * more than one faction throws outside production (see module docs). Use for
   * collections whose per-faction copies diverge (units), so callers are forced
   * to pass faction via {@link Collection.getInFaction} or opt out explicitly
   * with {@link Collection.getAny}. Requires {@link factionOf}.
   */
  guardUnscoped?: boolean;
  /**
   * Noun used in the {@link guardUnscoped} throw message (e.g. `"unit"`,
   * `"detachment"`). Defaults to `"entity"`. Cosmetic — steers the error toward
   * the right collection without changing behaviour.
   */
  entityLabel?: string;
  /** Wrap a raw record in its linked view. */
  wrap: (item: T) => V;
}

/**
 * A collection of one entity type, exposing id/name/faction lookups.
 *
 * Iterable: `for (const unit of units) { … }`.
 *
 * @typeParam T - the raw (generated) record type
 * @typeParam V - the linked view type returned to callers
 */
export class Collection<T, V> implements Iterable<V> {
  private readonly items: T[] = [];
  private readonly byId = new Map<string, T>();
  private readonly byNorm = new Map<string, T[]>();
  private readonly byFactionId = new Map<string, T[]>();
  private readonly byExternalRefKey = new Map<string, T[]>();
  private readonly idOf: (item: T) => string;
  private readonly nameOf?: (item: T) => string | undefined;
  private readonly idAliases?: Readonly<Record<string, string>>;
  private readonly wrapFn: (item: T) => V;
  /** Ids registered under >1 faction; only populated when guarding. */
  private readonly ambiguousIds?: Set<string>;
  /** Noun for the guard throw message. */
  private readonly entityLabel: string;

  constructor(cfg: CollectionConfig<T, V>) {
    this.idOf = cfg.idOf;
    this.nameOf = cfg.nameOf;
    this.idAliases = cfg.idAliases;
    this.wrapFn = cfg.wrap;
    this.entityLabel = cfg.entityLabel ?? "entity";
    const dedupeKeyOf = cfg.dedupeKeyOf ?? cfg.idOf;
    const seen = new Set<string>();
    // id -> distinct faction ids it appears under (only tracked when guarding).
    const idFactions = cfg.guardUnscoped
      ? new Map<string, Set<string>>()
      : undefined;
    for (const item of cfg.items) {
      const dedupeKey = dedupeKeyOf(item);
      if (seen.has(dedupeKey)) continue; // first-wins dedup
      seen.add(dedupeKey);
      this.items.push(item);

      const id = cfg.idOf(item);
      if (!this.byId.has(id)) this.byId.set(id, item); // first-wins for shared ids

      const name = cfg.nameOf?.(item);
      if (name) push(this.byNorm, normalizeName(name), item);

      // Alias names answer to the same record. Index them after the canonical
      // name and skip any alias that normalizes to an already-registered name,
      // so an alias can never displace the canonical owner of a normalized key.
      for (const alias of cfg.aliasesOf?.(item) ?? []) {
        const aliasKey = normalizeName(alias);
        if (
          aliasKey === "" ||
          aliasKey === (name ? normalizeName(name) : undefined)
        )
          continue;
        push(this.byNorm, aliasKey, item);
      }
      for (const ref of cfg.externalRefsOf?.(item) ?? []) {
        push(this.byExternalRefKey, externalRefKey(ref.namespace, ref.id), item);
      }

      const faction = cfg.factionOf?.(item);
      if (faction) {
        push(this.byFactionId, faction, item);
        if (idFactions) {
          const set = idFactions.get(id);
          if (set) set.add(faction);
          else idFactions.set(id, new Set([faction]));
        }
      }
    }
    if (idFactions) {
      this.ambiguousIds = new Set();
      for (const [id, factions] of idFactions) {
        if (factions.size > 1) this.ambiguousIds.add(id);
      }
    }
  }

  /** Every record, deduplicated by id, in first-seen order. */
  get all(): V[] {
    return this.items.map((item) => this.wrapFn(item));
  }

  /** Number of distinct records. */
  get size(): number {
    return this.items.length;
  }

  /**
   * Look up by exact id. For a guarded collection (see
   * {@link CollectionConfig.guardUnscoped}), an id that exists under more than
   * one faction throws outside production — pass a faction via
   * {@link getInFaction}, or call {@link getAny} when faction is genuinely
   * unknown. In production this degrades to first-wins.
   */
  get(id: string): V | undefined {
    if (!PRODUCTION && this.ambiguousIds?.has(id)) {
      throw new Error(
        `Ambiguous ${this.entityLabel} lookup: "${id}" exists under multiple factions; ` +
          `faction-less get() would return whichever copy registered first ` +
          `(wrong divergent fields). Use getInFaction("${id}", factionId), ` +
          `or getAny("${id}") when faction is genuinely unknown (import / conformance).`,
      );
    }
    const item = this.rawById(id);
    return item ? this.wrapFn(item) : undefined;
  }

  /**
   * Raw record for an id: exact `byId`, falling back to the {@link idAliases}
   * map (old id → current id) on a miss so a persisted reference to a renamed
   * id still resolves. Aliases are pre-flattened, so one hop is enough.
   */
  private rawById(id: string): T | undefined {
    const direct = this.byId.get(id);
    if (direct) return direct;
    const aliased = this.idAliases?.[id];
    return aliased !== undefined ? this.byId.get(aliased) : undefined;
  }

  /**
   * First-wins lookup by exact id that never throws, for callers with no
   * faction context on purpose (roster import, the conformance runner). For a
   * guarded collection this is the explicit opt-out of {@link get}'s ambiguity
   * tripwire; for an unguarded one it is identical to {@link get}.
   */
  getAny(id: string): V | undefined {
    const item = this.rawById(id);
    return item ? this.wrapFn(item) : undefined;
  }

  /**
   * Look up by exact id *within a faction*. Use this when an id is shared
   * across factions (e.g. `chaos-land-raider` lives under five Chaos factions)
   * and a faction context is known — {@link get} would return whichever copy
   * was registered first, which may belong to the wrong faction. Returns
   * `undefined` when no record with that id belongs to `factionId`.
   */
  getInFaction(id: string, factionId: string): V | undefined {
    // Resolve a renamed id to its current form before scoping to the faction.
    const resolvedId = this.byId.has(id) ? id : (this.idAliases?.[id] ?? id);
    const list = this.byFactionId.get(factionId);
    const item = list?.find((i) => this.idOf(i) === resolvedId);
    return item ? this.wrapFn(item) : undefined;
  }

  /** Whether a record with this exact id (or a renamed alias of it) exists. */
  has(id: string): boolean {
    return this.rawById(id) !== undefined;
  }

  /**
   * Return every canonical record carrying an exact external source identity.
   * External mappings are many-to-many: several records may share one source
   * identity, and one record may carry several ids from the same namespace.
   */
  byExternalRef(namespace: string, id: string): V[] {
    return (this.byExternalRefKey.get(externalRefKey(namespace, id)) ?? []).map(
      (item) => this.wrapFn(item),
    );
  }

  /**
   * Find one record by id or name. Name matching is diacritic- and
   * punctuation-insensitive (see {@link normalizeName}), trying, in order:
   * exact id → exact normalized name → normalized-name substring. Returns the
   * first match; names can repeat across factions, so use {@link findAll} or
   * {@link byFaction} when a query may be ambiguous.
   *
   * @example
   * units.find("Kharn"); // resolves "Khârn the Betrayer"
   */
  find(query: string): V | undefined {
    return this.findAll(query)[0];
  }

  /**
   * All records matching a query, by the same rules as {@link find}. An exact id
   * match returns just that record; otherwise every normalized-name-exact match
   * is returned, falling back to every normalized-name-substring match. Useful
   * to surface (rather than silently collapse) names shared across factions.
   */
  findAll(query: string): V[] {
    const byId = this.rawById(query);
    if (byId) return [this.wrapFn(byId)];

    const key = normalizeName(query);
    const exact = this.byNorm.get(key);
    if (exact && exact.length > 0) return exact.map((i) => this.wrapFn(i));

    if (!this.nameOf || key === "") return [];
    return this.items
      .filter((item) => normalizeName(this.nameOf!(item) ?? "").includes(key))
      .map((item) => this.wrapFn(item));
  }

  /** All records belonging to a faction id (empty if the type has no faction). */
  byFaction(factionId: string): V[] {
    return (this.byFactionId.get(factionId) ?? []).map((i) => this.wrapFn(i));
  }

  [Symbol.iterator](): Iterator<V> {
    return this.items.map((item) => this.wrapFn(item))[Symbol.iterator]();
  }
}

function push<K, T>(map: Map<K, T[]>, key: K, value: T): void {
  const existing = map.get(key);
  if (existing) existing.push(value);
  else map.set(key, [value]);
}

File: crates/wh40kdc/src/data/collection.rs
//! A queryable view over one entity collection.
//!
//! Indexes (by id, by normalized name, by faction) are built once at
//! construction. Records are deduplicated by a caller-supplied key (default:
//! id, first occurrence wins). Some records are intentionally shared: the same
//! id (e.g. unit `ministorum-priest`, ability `deadly-demise-d3`) appears under
//! several factions with per-faction copies that may diverge, so those
//! collections dedupe on `(faction_id, id)` to keep each faction's copy and
//! resolve via [`get_in_faction`](Collection::get_in_faction).
//!
//! [`find`](Collection::find) returns the first match when an id is shared
//! across factions; use [`by_faction`](Collection::by_faction) or
//! [`find_all`](Collection::find_all) to disambiguate. Collections whose
//! per-faction copies diverge arm [`with_unscoped_guard`](Collection::with_unscoped_guard)
//! so a faction-less [`get`](Collection::get) of a shared id panics in debug
//! builds; deliberately faction-less callers use [`get_any`](Collection::get_any).
//!
//! This mirrors the TypeScript `Collection` (`tools/src/data/collection.ts`);
//! unlike TS it stores owned records and hands back borrows (`&T`) rather than
//! lazily-wrapped view objects.

use crate::generated::ExternalReference;
use std::collections::HashMap;

use super::normalize::normalize_name;

/// A collection of one entity type, exposing id / name / faction lookups.
///
/// `T` is the raw (generated) record type. Built via [`Collection::build`],
/// which takes extractor closures but does not retain them — every index is
/// precomputed, so the struct holds no borrows and is free of self-referential
/// lifetimes.
pub struct Collection<T> {
    items: Vec<T>,
    by_id: HashMap<String, usize>,
    by_norm: HashMap<String, Vec<usize>>,
    by_faction: HashMap<String, Vec<usize>>,
    /// Per-faction id index: faction_id → (id → item index). Lets an id shared
    /// across factions resolve to a specific faction's copy (see `get_in_faction`).
    by_faction_id: HashMap<String, HashMap<String, usize>>,
    /// Exact external source identity → every matching item index.
    by_external_ref: HashMap<String, Vec<usize>>,
    /// Normalized name per item (parallel to `items`), for the substring fallback.
    norm_names: Vec<Option<String>>,
    /// Ids registered under >1 faction; populated by
    /// [`with_unscoped_guard`](Self::with_unscoped_guard).
    ambiguous_ids: Option<std::collections::HashSet<String>>,
    /// Renamed-id map (old id → current id), consulted by id lookups only on a
    /// `by_id` miss so a persisted reference to a since-renamed id still
    /// resolves. Pre-flattened, so one hop suffices. Populated by
    /// [`with_id_aliases`](Self::with_id_aliases). Mirror of the TS `idAliases`.
    id_aliases: Option<HashMap<String, String>>,
    /// Noun for the guard panic message (e.g. `"unit"`).
    entity_label: &'static str,
}

impl<T> Collection<T> {
    /// Build a collection, indexing each kept record by id, normalized name, and
    /// faction.
    ///
    /// - `id_of` — the record's primary id (e.g. `|u| u.id.to_string()`).
    ///   Returns an owned `String` so string-newtype ids (`EntityId`) and
    ///   `Display`-based string enums (e.g. `ForceDispositionId`) work alike;
    ///   it is only invoked at build time.
    /// - `name_of` — its display name, if any (drives [`find`](Self::find)).
    /// - `faction_of` — its owning faction id, if any (drives
    ///   [`by_faction`](Self::by_faction)).
    /// - `dedupe_of` — the uniqueness key; first occurrence wins. Pass a
    ///   composite (e.g. `(faction_id, id)`) for records that legitimately share
    ///   an id across factions so distinct copies are preserved.
    pub fn build(
        items: Vec<T>,
        id_of: impl Fn(&T) -> String,
        name_of: impl Fn(&T) -> Option<&str>,
        faction_of: impl Fn(&T) -> Option<&str>,
        dedupe_of: impl Fn(&T) -> String,
    ) -> Self {
        // No aliases by default — only some collections (units) answer to them.
        Self::build_with_aliases(items, id_of, name_of, |_| Vec::new(), faction_of, dedupe_of)
    }

    /// Like [`build`](Self::build) but also indexes alternate names.
    ///
    /// - `aliases_of` — alternate names the record answers to (e.g. spelling
    ///   variants from other tools' exports). Each is indexed into the
    ///   normalized-name map alongside the canonical name, so
    ///   [`find`](Self::find) / [`find_all`](Self::find_all) match an alias
    ///   exactly — but an alias is never returned as the record's display name,
    ///   and the canonical name always wins a collision (aliases are appended
    ///   after the canonical entry and any alias that normalizes to the record's
    ///   own canonical name is skipped). Mirror of the TS `aliasesOf`.
    pub fn build_with_aliases(
        items: Vec<T>,
        id_of: impl Fn(&T) -> String,
        name_of: impl Fn(&T) -> Option<&str>,
        aliases_of: impl Fn(&T) -> Vec<&str>,
        faction_of: impl Fn(&T) -> Option<&str>,
        dedupe_of: impl Fn(&T) -> String,
    ) -> Self {
        let mut kept: Vec<T> = Vec::with_capacity(items.len());
        let mut by_id: HashMap<String, usize> = HashMap::new();
        let mut by_norm: HashMap<String, Vec<usize>> = HashMap::new();
        let mut by_faction: HashMap<String, Vec<usize>> = HashMap::new();
        let mut by_faction_id: HashMap<String, HashMap<String, usize>> = HashMap::new();
        let mut norm_names: Vec<Option<String>> = Vec::with_capacity(items.len());
        let mut seen: std::collections::HashSet<String> = std::collections::HashSet::new();

        for item in items {
            if !seen.insert(dedupe_of(&item)) {
                continue; // first-wins dedup
            }
            let idx = kept.len();

            // First-wins for shared ids (a later faction copy doesn't clobber).
            by_id.entry(id_of(&item)).or_insert(idx);

            let norm = name_of(&item).map(normalize_name);
            if let Some(key) = &norm {
                by_norm.entry(key.clone()).or_default().push(idx);
            }
            // Index aliases after the canonical name, skipping empties and any
            // alias that normalizes to the canonical name, so an alias can never
            // displace the canonical owner of a normalized key.
            for alias in aliases_of(&item) {
                let alias_key = normalize_name(alias);
                if alias_key.is_empty() || norm.as_deref() == Some(alias_key.as_str()) {
                    continue;
                }
                by_norm.entry(alias_key).or_default().push(idx);
            }
            norm_names.push(norm);

            if let Some(faction) = faction_of(&item) {
                by_faction.entry(faction.to_string()).or_default().push(idx);
                by_faction_id
                    .entry(faction.to_string())
                    .or_default()
                    .entry(id_of(&item))
                    .or_insert(idx);
            }

            kept.push(item);
        }

        Self {
            items: kept,
            by_id,
            by_norm,
            by_faction,
            by_faction_id,
            by_external_ref: HashMap::new(),
            norm_names,
            ambiguous_ids: None,
            id_aliases: None,
            entity_label: "entity",
        }
    }

    /// Arm the unscoped-lookup tripwire: after this, a faction-less
    /// [`get`](Self::get) of an id that exists under more than one faction
    /// panics in debug builds (it would silently return whichever copy
    /// registered first — the wrong divergent fields). Use for collections
    /// whose per-faction copies diverge (units, detachments, weapons,
    /// abilities), so callers are forced through
    /// [`get_in_faction`](Self::get_in_faction) or opt out explicitly with
    /// [`get_any`](Self::get_any). Release builds degrade to the historical
    /// first-wins behaviour rather than crashing a consumer. Mirror of the TS
    /// `guardUnscoped` (whose non-production tripwire maps to
    /// `debug_assertions` here).
    pub fn with_unscoped_guard(mut self, entity_label: &'static str) -> Self {
        let mut ambiguous: std::collections::HashSet<String> = std::collections::HashSet::new();
        let mut seen: HashMap<&str, u32> = HashMap::new();
        for per_faction in self.by_faction_id.values() {
            for id in per_faction.keys() {
                *seen.entry(id.as_str()).or_insert(0) += 1;
            }
        }
        for (id, count) in seen {
            if count > 1 {
                ambiguous.insert(id.to_string());
            }
        }
        self.ambiguous_ids = Some(ambiguous);
        self.entity_label = entity_label;
        self
    }

    /// Attach a renamed-id map (old id → current id). Id lookups
    /// ([`get`](Self::get)/[`get_any`](Self::get_any)/[`get_in_faction`](Self::get_in_faction)/
    /// [`has`](Self::has)) consult it only when the exact id misses, so a
    /// persisted reference to a since-renamed id still resolves. The map is
    /// expected to be pre-flattened (each key maps directly to a terminal live
    /// id). Mirror of the TS `idAliases` (typically the share registry's
    /// `aliases`).
    pub fn with_id_aliases(mut self, id_aliases: HashMap<String, String>) -> Self {
        self.id_aliases = Some(id_aliases);
        self
    }

    /// Build the many-to-many external-source identity index.
    pub fn with_external_refs(mut self, refs_of: impl Fn(&T) -> &[ExternalReference]) -> Self {
        for (index, item) in self.items.iter().enumerate() {
            for reference in refs_of(item) {
                let key = format!(
                    "{}\0{}",
                    reference.namespace.as_str(),
                    reference.id.as_str()
                );
                self.by_external_ref.entry(key).or_default().push(index);
            }
        }
        self
    }

    /// Item index for an id: exact `by_id`, falling back through the
    /// [`id_aliases`](Self::with_id_aliases) map (one hop) on a miss.
    fn raw_index(&self, id: &str) -> Option<usize> {
        if let Some(&i) = self.by_id.get(id) {
            return Some(i);
        }
        self.id_aliases
            .as_ref()
            .and_then(|m| m.get(id))
            .and_then(|new_id| self.by_id.get(new_id).copied())
    }

    /// Every record, deduplicated, in first-seen order.
    pub fn all(&self) -> &[T] {
        &self.items
    }

    /// Number of distinct records.
    pub fn len(&self) -> usize {
        self.items.len()
    }

    /// Whether the collection holds no records.
    pub fn is_empty(&self) -> bool {
        self.items.is_empty()
    }

    /// Return every record carrying the exact external source identity.
    ///
    /// External mappings are many-to-many, so this always returns a collection.
    pub fn by_external_ref(&self, namespace: &str, id: &str) -> Vec<&T> {
        let key = format!("{namespace}\0{id}");
        self.by_external_ref
            .get(&key)
            .map(|indices| indices.iter().map(|&index| &self.items[index]).collect())
            .unwrap_or_default()
    }

    /// Look up by exact id.
    ///
    /// For a guarded collection (see
    /// [`with_unscoped_guard`](Self::with_unscoped_guard)), an id that exists
    /// under more than one faction panics in debug builds — pass a faction via
    /// [`get_in_faction`](Self::get_in_faction), or call
    /// [`get_any`](Self::get_any) when faction is genuinely unknown. Release
    /// builds degrade to first-wins.
    pub fn get(&self, id: &str) -> Option<&T> {
        #[cfg(debug_assertions)]
        if let Some(ambiguous) = &self.ambiguous_ids {
            if ambiguous.contains(id) {
                panic!(
                    "Ambiguous {} lookup: \"{id}\" exists under multiple factions; \
                     a faction-less get() would return whichever copy registered first \
                     (wrong divergent fields). Use get_in_faction(\"{id}\", faction_id), \
                     or get_any(\"{id}\") when faction is genuinely unknown (import / conformance).",
                    self.entity_label,
                );
            }
        }
        self.raw_index(id).map(|i| &self.items[i])
    }

    /// First-wins lookup by exact id that never panics, for callers with no
    /// faction context on purpose (roster import, the conformance runner). For
    /// a guarded collection this is the explicit opt-out of
    /// [`get`](Self::get)'s ambiguity tripwire; for an unguarded one it is
    /// identical to `get`. Mirror of the TS `getAny`.
    pub fn get_any(&self, id: &str) -> Option<&T> {
        self.raw_index(id).map(|i| &self.items[i])
    }

    /// Look up by exact id *within a faction*. Returns the record with `id`
    /// belonging to `faction_id`, or `None`. Use when an id is shared across
    /// factions and faction context is known — [`get`](Self::get) returns
    /// whichever copy registered first, which may be the wrong faction's. Mirror
    /// of the TS `getInFaction`.
    pub fn get_in_faction(&self, id: &str, faction_id: &str) -> Option<&T> {
        // Resolve a renamed id to its current form before scoping to the faction.
        let resolved: &str = if self.by_id.contains_key(id) {
            id
        } else {
            self.id_aliases
                .as_ref()
                .and_then(|m| m.get(id))
                .map(String::as_str)
                .unwrap_or(id)
        };
        self.by_faction_id
            .get(faction_id)
            .and_then(|m| m.get(resolved))
            .map(|&i| &self.items[i])
    }

    /// Whether a record with this exact id (or a renamed alias of it) exists.
    pub fn has(&self, id: &str) -> bool {
        self.raw_index(id).is_some()
    }

    /// Record at a stored index (used by [`Dataset`](super::Dataset)'s reverse
    /// indexes, which hold `usize` positions into this collection).
    pub(super) fn at(&self, idx: usize) -> &T {
        &self.items[idx]
    }

    /// Find one record by id or name, returning the first match.
    ///
    /// Name matching is diacritic- and punctuation-insensitive (see
    /// [`normalize_name`](super::normalize_name)), trying in order: exact id →
    /// exact normalized name → normalized-name substring. Names can repeat
    /// across factions, so use [`find_all`](Self::find_all) or
    /// [`by_faction`](Self::by_faction) when a query may be ambiguous.
    pub fn find(&self, query: &str) -> Option<&T> {
        self.find_all(query).into_iter().next()
    }

    /// All records matching a query, by the same rules as [`find`](Self::find).
    ///
    /// An exact id match returns just that record; otherwise every
    /// normalized-name-exact match is returned, falling back to every
    /// normalized-name-substring match. Surfaces (rather than silently
    /// collapses) names shared across factions.
    pub fn find_all(&self, query: &str) -> Vec<&T> {
        if let Some(i) = self.raw_index(query) {
            return vec![&self.items[i]];
        }
        let key = normalize_name(query);
        if let Some(idxs) = self.by_norm.get(&key) {
            if !idxs.is_empty() {
                return idxs.iter().map(|&i| &self.items[i]).collect();
            }
        }
        if key.is_empty() {
            return Vec::new();
        }
        self.norm_names
            .iter()
            .enumerate()
            .filter_map(|(i, n)| match n {
                Some(name) if name.contains(&key) => Some(&self.items[i]),
                _ => None,
            })
            .collect()
    }

    /// All records belonging to a faction id (empty if the type has no faction).
    pub fn by_faction(&self, faction_id: &str) -> Vec<&T> {
        self.by_faction
            .get(faction_id)
            .map(|idxs| idxs.iter().map(|&i| &self.items[i]).collect())
            .unwrap_or_default()
    }

    /// Iterate the distinct records in first-seen order.
    pub fn iter(&self) -> std::slice::Iter<'_, T> {
        self.items.iter()
    }
}

impl<'a, T> IntoIterator for &'a Collection<T> {
    type Item = &'a T;
    type IntoIter = std::slice::Iter<'a, T>;

    fn into_iter(self) -> Self::IntoIter {
        self.items.iter()
    }
}
